anaconda/pyanaconda/bootloader.py

```diff
@@ -1908,16 +1908,166 @@ def is_valid_stage2_device(self, device, linux=True, non_linux=False):
 
     write_config = BootLoader.write_config
 
-class MacEFIGRUB(EFIGRUB):
+class MacEFIGRUB(XenEFI):
+    """Special EFI handling for macOS HFS+ ESP partition.
+
+    Typical GRUB2 installations would execute the script
+    located at /usr/libexec/mactel-boot-setup which would
+    modify the HFS+ ESP files and bless the specified efi.
+
+    However, we are not using GRUB at this time which would
+    cause that script to exit earlier.  
+
+    In this class, we will execute the relevant commands
+    to symlink the efi file in the /System directory as well
+    the cfg file.  Lastly, macOS requires the bootable efi
+    file to be blessed.
+    """
+
     def mactel_config(self):
-        if os.path.exists(iutil.getSysroot() + "/usr/libexec/mactel-boot-setup"):
-            rc = iutil.execInSysroot("/usr/libexec/mactel-boot-setup", [])
-            if rc:
-                log.error("failed to configure Mac boot loader")
+        """Modifies the HFS+ ESP partition to be bootable.
+
+        Based on the /usr/libexec/mactel-boot-setup script, 
+        we will create two symlinks pointing to the Qubes
+        Xen version that is installed and configured for
+        the system.  We also need to run hfs-bless on
+        the specific EFI file we allow to be bootable.
+        """
+
+        log.info("Configuring mactel Mac EFI")
+
+        efi_fil = "{}/{}".format(self.config_dir, "xen.efi")
+        efi_cfg = "{}/{}".format(self.config_dir, "xen.cfg")
+        sys_dir = "/boot/efi/System/Library/Coreservices"
+
+        not_ready = False
+        if not os.path.exists(iutil.getSysroot() + efi_fil):
+            log.error("mactel efi_fil not found: %s", iutil.getSysroot() + efi_fil)
+            not_ready = True
+
+        if not os.path.exists(iutil.getSysroot() + efi_cfg):
+            log.error("mactel efi_cfg not found: %s", iutil.getSysroot() + efi_cfg)
+            not_ready = True
+
+        if not_ready:
+            log.error("mactel cannot continue with mactel_config. see log for details.")
+            return
+
+        root_sys_dir = iutil.getSysroot() + sys_dir
+        if not os.path.exists(root_sys_dir):
+            try:
+                os.makedirs(root_sys_dir)
+            except:
+                log.warning("mactel tried to create root_sys_dir %s but received error", root_sys_dir)
+        
+        sys_efi = "{}/{}".format(sys_dir, "boot.efi")
+        rc = iutil.execInSysroot("ln", ["-f", efi_fil, sys_efi])
```

Comment:
Why hardlink for xen.efi, but symlink for xen.cfg? I guess it didn't worked for you with symlink. But then why symlink below isn't a problem?

Note that /boot/efi is a separate partition, and may be mounted on different path during system boot (I have no idea how Apple EFI handle it). So, symlinks using absolute path may appear to be broken in some cases. If using symlinks (which IMO is better idea here), then relative path is safer. In this case it would be `"../../../EFI/{}/xen.efi".format(self.efi_dir)`.

Why hardlink or symlink: subsequent updates (Xen, kernel etc) will update xen.efi (and/or xen.cfg) in its original path. If we use hardlink here, it may be broken during the update, and old version of Xen (or kernel) will still be used. This may be security issue. On the other hand, if using symlink, there is only one place where the file reside. If will be moved/renamed/deleted, symlink will be broken. This probably will lead to broken system, but it is better (especially in system like Qubes OS) than false sense of security (Xen package updated, but in reality old version is used).

Reply:
Also note that `mactel-boot` package ship `boot.efi` file, which I think is GRUB. Since you bless `xen.efi`, not `boot.efi`, maybe it is enough to simply remove `boot.efi`? To avoid time consuming reinstallation, you can try on installed system - mount that partition and remove `boot.efi` and `boot.cfg`. But if it doesn't work, try with relative symlinks as in comment above (again - you may want do it manually first, before trying reinstallation).

Reply:
> Why hardlink for xen.efi, but symlink for xen.cfg? 
I followed what all other Mac EFI bootloaders perform (under grub), and that is hardlink the EFI file and softlink the CFG file.  I actually haven't tested trying softlinks for both.  

Obviously, we do what some type of link so not to "duplicate" the files - and to allow for future upgrades of the standard location.

I will test it under softlinks and see how it works.

> Note that /boot/efi is a separate partition, and may be mounted on different path during system boot (I have no idea how Apple EFI handle it). So, symlinks using absolute path may appear to be broken in some cases. If using symlinks (which IMO is better idea here), then relative path is safer. In this case it would be `"../../../EFI/{}/xen.efi".format(self.efi_dir)`.
>
> Why hardlink or symlink: subsequent updates (Xen, kernel etc) will update xen.efi (and/or xen.cfg) in its original path. If we use hardlink here, it may be broken during the update, and old version of Xen (or kernel) will still be used. This may be security issue. On the other hand, if using symlink, there is only one place where the file reside. If will be moved/renamed/deleted, symlink will be broken. This probably will lead to broken system, but it is better (especially in system like Qubes OS) than false sense of security (Xen package updated, but in reality old version is used).

I completely agree as I did not want to hardcode paths either.  I tried `os.symlink` with relative paths, but since it runs outside of `chroot` it didn't work either. 

I will attempt to use relative paths with the `execInSysroot` method I am using now.

> Also note that mactel-boot package ship boot.efi file, which I think is GRUB.

I'd like to remove this package at some point.  But since it installs the `hfs-bless` binary (which is a separate tar within the package), I am not sure that we can anytime soon without writing a whole new package.

> Since you bless xen.efi, not boot.efi, maybe it is enough to simply remove boot.efi? To avoid time consuming reinstallation, you can try on installed system - mount that partition and remove boot.efi and boot.cfg. But if it doesn't work, try with relative symlinks as in comment above (again - you may want do it manually first, before trying reinstallation).

The boot.* files located at `/boot/efi/System/Library/Coreservices` are required for Apple to pick them up for the "Boot Camp" and "Startup Disk" macOS applications and the macOS `bless` CLI tool for switching and setting defaults for booting.  This is something I rigorously tested on my systems, since there is no public documentation about it.  They have to be there.

I will switch to a symlink/softlink and test, yep manually first.

I will also attempt to remove all hardcoded.  


Ps, this is a lot easier in bash scripting - like the `mactel.boot` package originally had.  :)

Reply:
> Ps, this is a lot easier in bash scripting - like the mactel.boot package originally had. :)

You can also do that from script. There is "anaconda/scripts" directory and you can add new script there. And add to `Makefile.am` to install it (for example to `dist_stage2scripts_SCRIPTS` - will get installed into `/usr/share/anaconda/...`). Then call that script from this function.

BTW I've just learned there is `/usr/share/anaconda/restart-anaconda` - maybe useful for debugging.